mono/metadata/threads.c

```diff
@@ -1599,38 +1599,33 @@ ves_icall_System_Threading_InternalThread_Thread_free_internal (MonoInternalThre
 }
 
 void
-ves_icall_System_Threading_Thread_Sleep_internal (gint32 ms)
+ves_icall_System_Threading_Thread_Sleep_internal (gint32 ms, MonoError *error)
 {
-	guint32 res;
-	MonoInternalThread *thread = mono_thread_internal_current ();
-
 	THREAD_DEBUG (g_message ("%s: Sleeping for %d ms", __func__, ms));
 
 	if (mono_thread_current_check_pending_interrupt ())
 		return;
 
+	MonoInternalThread * const thread = mono_thread_internal_current ();
+
 	while (TRUE) {
 		gboolean alerted = FALSE;
 
 		mono_thread_set_state (thread, ThreadState_WaitSleepJoin);
 
-		res = mono_thread_info_sleep (ms, &alerted);
+		(void)mono_thread_info_sleep (ms, &alerted);
 
 		mono_thread_clr_state (thread, ThreadState_WaitSleepJoin);
 
-		if (alerted) {
-			MonoException* exc = mono_thread_execute_interruption_ptr ();
-			if (exc) {
-				mono_set_pending_exception (exc);
-				return;
-			} else {
-				// FIXME: !MONO_INFINITE_WAIT
-				if (ms != MONO_INFINITE_WAIT)
-					break;
-			}
-		} else {
-			break;
-		}
+		if (!alerted)
+			return;
+
+		MonoExceptionHandle exc;
+		if (mono_thread_execute_interruption (&exc))
```

Comment:
There are two ways this can play out:
* if `mono_thread_execute_interruption` will write a to `&exc` a handle that's created within a `HANDLE_FUNCTION_ENTER/HANDLE_FUNCTION_RETURN` pair, then `exc` will be a handle that's already deallocated
* if `mono_thread_execute_interruption` writes a handle that's allocated outside a `HANDLE_FUNCTION_ENTER/HANDLE_FUNCTION_RETURN` pair, then each loop iteration will allocate a new handle and overall we will allocate an unbounded number of handles, which is not allowed.

One thing that could work instead is:
```c
{
  // allocate a handle once outside the loop body
  MonoExceptionHandle exc = MONO_HANDLE_NEW (MonoException, NULL);
  while (TRUE) {
     ...
     // expect callee to assign to exc if it returns TRUE
     if (mono_thread_execute_interruption (exc)) {
        ... // do something with exc here
        break; 
     }
  }
}
...
gboolean
mono_thread_execute_interruption (MonoExceptionHandleOut excOut) {
  HANDLE_FUNCTION_ENTER ();
  gboolean success;
  ...
  MonoExceptionHandle excLocal = MONO_HANDLE_NEW (MonoException, ...);
  ...
  MONO_HANDLE_ASSIGN (excOut, excLocal); // make excOut point to what excLocal points to
  HANDLE_FUNCTION_RETURN (success);
}
```
Where `MonoExceptionHandleOut` is just another typedef for `MonoExceptionHandle` (it automatically gets created by the handle.h macros as the same time as `MonoExceptionHandle`)

Reply:
1. I didn't think deallocation worked that way -- it is giving the caller his own thread local.
2. The callers do not generally or at al loop, once they get a handle. Previously we'd create null . handles, and had to move that out of the loop. Now we only create non-null handles so the looping is really gone. Maybe fragile, but gone.

Reply:
I don't use hungarian either but I need three forms of exc, fexc and pexc. i could also call it exc_out and exc_present or exc_bool, though exc_bool is just verbose Hungarian.

Reply:
The out parameter cannot be outside of a enter/exit pair. Creators of local handles, that are then address taken for out parameters, know to have a frame to contain them.

Reply:
1. handles are per-thread, but they're also allocated in regions: `HANDLE_FUNCTION_ENTER` begins a region and `HANDLE_FUNCTION_RETURN` leaves the region and deallocates all the handles that were allocated inside.

2. I missed that that the loop runs only while `exc` hasn't been written to, sorry.  I think that warrants a sizeable comment before `mono_thread_execute_interruption` that describes how to use it properly.  Or a simpler design.

3. yes I missed a handle enter/return pair in the top half of my example.
   ```c
      {
        HANDLE_FUNCTION_ENTER();
        MonoExceptionHandle exc = MONO_HANDLE_NEW (MonoException, NULL);
        while (TRUE)  {
          ...
          if (mono_thread_execute_interruption (exc)) {
            ... // use exc;
            break;
        }
        HANDLE_FUNCTION_RETURN();
   }
   ```